Sources/MapboxDirections/RouteRoadClassesViolations.swift

```diff
@@ -0,0 +1,69 @@
+
+import Foundation
+
+/**
+ Description of `RoadClasses` that were meant to be avoided during routing but still got used in a `Route`.
+ 
+ Such ignoring may be done by the engine if it cannot provide a route otherwise.
+ */
+public struct RouteRoadClassesViolations {
+    /**
+     Related `Route` object
+     */
+    public let route: Route
+    
+    /**
+     List of all existing violations, down to `RouteStep` granularity.
+     */
+    public let violations: [RoadClassExclusionViolation]
+    
+    /**
+     Filters `violations` to search for specific leg, step and intersection.
+     
+     - parameter legIndex: Index of a leg inside related `Route`to search in.
+     - parameter stepIndex: Index of a step inside given `Route`'s leg.
+     - parameter intersectionIndex: Index of an intersection inside given `Route`'s leg and step.
+     - returns: Array of `RoadClassExclusionViolation` under given indicies.
+     
+     Passing `nil` as one of the indicies will result in searching for all violations at given `legIndex`/`stepIndex`.
+     - note: Searching by `intersectionIndex` requires non-nil `stepIndex`.
+     */
+    public func violations(at legIndex: Int, stepIndex: Int? = nil, intersectionIndex: Int? = nil) -> [RoadClassExclusionViolation] {
+        assert(!(stepIndex == nil && intersectionIndex != nil), "It is forbidden to select `intersectionIndex` without specifying `stepIndex`.")
```

Comment:
The invariant can be embedded in the type-system. For example by using two methods:
```swift
public func violations(at legIndex: Int, stepIndex: Int, intersectionIndex: Int) -> [RoadClassExclusionViolation]
```

```swift
public func violations(at legIndex: Int, stepIndex: Int? = nil) -> [RoadClassExclusionViolation]
```

That way it will be impossible for developers to misuse the method and simplify the methods.

Reply:
Good idea. Code updated